### DISCUSSION TITLE
How to use a subclass of BaseModel like a dictionary

### DISCUSSION THREAD

**Discussion by ericchansen** [pydantic/pydantic#4069] [Question]
I need to be able to subclass `BaseModel` such that it can be interacted with like a dictionary.

Should I define `__iter__`, which overwrites the `__iter__` in `BaseModel`, or should I define `items`? You only need one of the two for everything to work.

```
import sys
from typing import Any

from pydantic import BaseModel

if sys.version_info[:2] >= (3, 8):
    from collections.abc import MutableMapping
else:
    from collections import MutableMapping


class DictLikeBaseModel(BaseModel, MutableMapping):
    a: str
    b: str
    c: str

    def __contains__(self, x: str) -> bool:
        return True if x in self.__dict__.keys() else False

    def __delitem__(self, x: str) -> None:
        del self.__dict__[x]

    def __getitem__(self, x: str) -> Any:
        return self.__dict__[x]

    # Should I define __iter__ (which overwrites the __iter__ in BasesModel)...
    def __iter__(self):
        return iter(self.__dict__)

    def __len__(self) -> int:
        return len(self.__dict__)

    def __setitem__(self, key: str, value: Any) -> None:
        self.__dict__[key] = value

    # ... or this?
    def items(self):
        return self.__dict__.items()


x = DictLikeBaseModel(a="a", b="b", c="c")
print(len(x))
print(x["a"])
x["a"] = "wow"
for k, v in x.items():
    print(k, v)
```

**Comment by micjulia** [comment]
Is there a reason you can't just use the `dict` method? It's unclear to me what the benefit of using the model class is after it has validated everything if you really just need a dictionary.

**Reply by ericchansen** [reply]
I just realized that validation doesn't happen after instantiation (see example below).

```
from pydantic import BaseModel, PositiveInt


class AnotherModel(BaseModel):
    a: str
    b: PositiveInt


x = AnotherModel(a="a", b=1)
print(f"b: {x.b}")
x.b = -1
print(f"b: {x.b}")
```

That's a bummer. I guess we could just pass a dict around. I was thinking that attempting to do `x.b = -1` would raise a `ValidationError`.

**Reply by ericchansen** [reply]
One motivation for the usefulness might be to still have access to the aliases. Once you convert to a dictionary, you lose the aliases.

Say you need to be able to grab an attribute based on the field name or the alias. Something like this is possible (and there are probably more graceful implementations of this).

```
from pydantic import BaseModel, Field


class AnotherModel(BaseModel):
    a: str
    b: str = Field(..., alias="wow")


x = AnotherModel(a="a", wow="wow")
print(f"b: {x.b}")
print(f"wow: {x.dict(by_alias=True)['wow']}")
```

**Reply by jnawk** [reply]
It would be nice to be able to treat the model subclass like a dictionary, to enable backwards compatibility with code that used to be fed a dict but is now being fed a model subclass - the problem I'm trying to solve is where the old code is doing stuff like json.dumps with what it thinks is a dictionary - indeed, if isinstance(model, dict) was able to return true, then json.dumps would work just fine.
